File: Data Structures & Algorithms/search-for-word-ii/submission-5.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False
        self.word = None
    
class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            
            node = node.children[ch]
        
        node.isEnd = True
        node.word = word

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        for word in words:
            trie.insert(word)
        
        m, n = len(board), len(board[0])
        res = set()

        def dfs(row, col, node):
            ch = board[row][col]
            child = node.children.get(ch)
            if not child:
                return
            
            if child.isEnd:
                res.add(child.word)
            
            board[row][col] = "#"
            for dx, dy in [(-1, 0), (1, 0), (0, 1), (0, -1)]:
                nx, ny = dx + row, dy + col
                if 0 <= nx < m and 0 <= ny < n and board[nx][ny] != "#":
                    dfs(nx, ny, child)
            
            board[row][col] = ch
            
        for i in range(m):
            for j in range(n):
                if board[i][j] in trie.root.children:
                    dfs(i, j, trie.root)
        
        return list(res)

File: Data Structures & Algorithms/n-queens/submission-5.java
class Solution {
    private List<List<String>> res = new ArrayList<>();
    private char[][] board;

    public List<List<String>> solveNQueens(int n) {
        board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        backtrack(board, 0);
        return res;
    }

    private void backtrack(char[][] board, int row) {
        if (row == board.length) {
            List<String> solution = new ArrayList<>();
            for (char[] r : board) {
                solution.add(new String(r));
            }
            res.add(solution);
            return;
        }

        for (int col = 0; col < board[row].length; col++) {
            if (!isValue(board, row, col)) {
                continue;
            }
            board[row][col] = 'Q';
            backtrack(board, row+1);
            board[row][col] = '.';
        }
    }

    private boolean isValue(char[][] board, int row, int col) {
        int n = board.length;
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }


        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }
}
